tests/kernel/futex/src/main.c

```diff
@@ -0,0 +1,379 @@
+K_APPMEM_PARTITION_DEFINE(futex_partition);
+K_APP_BMEM(futex_partition) int timeout;
+K_APP_BMEM(futex_partition) int index[TOTAL_THREADS_WAITING];
+K_APP_BMEM(futex_partition) struct k_futex simple_futex;
+K_APP_BMEM(futex_partition) struct k_futex multiple_futex[TOTAL_THREADS_WAITING];
+
+struct k_thread futex_tid;
+struct k_thread multiple_tid[TOTAL_THREADS_WAITING];
+
+/******************************************************************************/
+/* Helper functions */
+void futex_isr_wake(void *futex)
+{
+	k_futex_wake((struct k_futex *)futex, 1);
+}
+
+void futex_wait_task(void *p1, void *p2, void *p3)
+{
+	s32_t ret_value;
+	int time_val = *(int *)p1;
+
+	zassert_true(time_val >= K_FOREVER, "invalid timeout parameter");
+
+	ret_value = k_futex_wait(&simple_futex,
+			atomic_get(&simple_futex.val), time_val);
+
+	switch (time_val) {
+	case K_FOREVER:
+		zassert_true(ret_value == 0,
+		     "k_futex_wait failed when it shouldn't have");
+		zassert_false(ret_value == 0,
+		     "futex_wait_task wakeup when it shouldnt't have");
+		break;
+	case K_NO_WAIT:
+		zassert_true(ret_value == -EBUSY,
+		     "k_futex_wait failed when it shouldn't have");
+		atomic_sub(&simple_futex.val, 1);
+		break;
+	default:
+		zassert_true(ret_value == -ETIMEDOUT,
+		     "k_futex_wait failed when it shouldn't have");
+		atomic_sub(&simple_futex.val, 1);
+		break;
+	}
+}
+
+void futex_wait_wake_task(void *p1, void *p2, void *p3)
+{
+	s32_t ret_value;
+	int time_val = *(int *)p1;
+
+	zassert_true(time_val >= K_FOREVER, "invalid timeout parameter");
+
+	ret_value = k_futex_wait(&simple_futex,
+			atomic_get(&simple_futex.val), time_val);
+
+	switch (time_val) {
+	case K_FOREVER:
+		zassert_true(ret_value == 0,
+		     "k_futex_wait failed when it shouldn't have");
+		break;
+	case K_NO_WAIT:
+		zassert_true(ret_value == -EBUSY,
+		     "k_futex_wait failed when it shouldn't have");
+		break;
+	default:
+		zassert_true(ret_value == 0,
+		     "k_futex_wait failed when it shouldn't have");
+		break;
+	}
+
+	atomic_sub(&simple_futex.val, 1);
+}
+
+void futex_multiple_wait_wake_task(void *p1, void *p2, void *p3)
+{
+	s32_t ret_value;
+	int time_val = *(int *)p1;
+	int idx = *(int *)p2;
+
+	zassert_true(time_val == K_FOREVER, "invalid timeout parameter");
+
+	ret_value = k_futex_wait(&multiple_futex[idx],
+		atomic_get(&(multiple_futex[idx].val)), time_val);
+	zassert_true(ret_value == 0,
+	     "k_futex_wait failed when it shouldn't have");
+
+	atomic_sub(&(multiple_futex[idx].val), 1);
+}
+
+/**
+ * @ingroup futex_tests
+ * @{
+ */
+
+/**
+ * @brief Test k_futex_wait() forever
+ */
+void test_futex_wait_forever(void)
+{
+	timeout = K_FOREVER;
+
+	k_futex_reset(&simple_futex);
+	atomic_inc(&simple_futex.val);
+
+	k_thread_create(&futex_tid, stack_1, STACK_SIZE,
+			futex_wait_task, &timeout, NULL, NULL,
+			K_PRIO_PREEMPT(0), K_USER | K_INHERIT_PERMS,
+			K_NO_WAIT);
+
+	/* giving time for the futex_wait_task to execute */
+	k_sleep(K_MSEC(200));
```

Comment:
let's not add sleeps like this. Please note that CONFIG_ZTEST_THREAD_PRIORITY is of the cooperative priority of -1. creating the child thread at the same or higher priority, and then calling k_yield() should do the same thing. you should be able to get rid of most of the k_sleep() calls in this C file.

Reply:
yes, this is a good idea to make deterministic output. Thanks